tools/mpremote/mpremote/pyboardextended.py

```diff
@@ -621,37 +621,66 @@ def mount_local(self, path):
         self.cmd = PyboardCommand(self.serial, fout, path)
         self.serial = SerialIntercept(self.serial, self.cmd)
 
-    def soft_reset_with_mount(self, out_callback):
+    def write_ctrl_d(self, out_callback):
         self.serial.write(b"\x04")
         if not self.mounted:
             return
 
-        # Clear flag while board reboots, it will be re-set once mounted.
-        self.mounted = False
+        # Read response from the device until it is quiet (with a timeout).
+        INITIAL_TIMEOUT = 0.5
+        QUIET_TIMEOUT = 0.2
+        FULL_TIMEOUT = 5
+        t_start = t_last_activity = time.monotonic()
+        data_all = b""
+        while True:
+            t = time.monotonic()
+            n = self.serial.inWaiting()
+            if n > 0:
+                data = self.serial.read(n)
+                out_callback(data)
+                data_all += data
+                t_last_activity = t
+            else:
+                if len(data_all) == 0:
+                    if t - t_start > INITIAL_TIMEOUT:
+                        return
+                else:
+                    if t - t_start > FULL_TIMEOUT:
+                        return
+                    if t - t_last_activity > QUIET_TIMEOUT:
+                        break
 
-        # Wait for a response to the soft-reset command.
-        for i in range(10):
-            if self.serial.inWaiting():
-                break
-            time.sleep(0.05)
+        # Check if a soft reset occurred.
+        if data_all.find(b"MPY: soft reboot") == -1:
```

Comment:
Pyserial has some functions that could make this simpler, eg 
`read_until()` reads input while watching for a desired pattern or timeout (
https://pyserial.readthedocs.io/en/latest/pyserial_api.html#serial.Serial.read_until)

Or even just using its built in timeouts rather than manually managing it here?
It's got separate settings for overall read operation timeout (https://pyserial.readthedocs.io/en/latest/pyserial_api.html#serial.Serial.timeout) and inter byte timeouts (https://pyserial.readthedocs.io/en/latest/pyserial_api.html#serial.Serial.inter_byte_timeout)

Edit: 
Ah, I see there's a custom pyboard.read_until that could possibly be used, but doesn't have as fine grain timeout second. Also, the compatibility layer with telnet doesn't have the extra timeout settings, that's one reason to not use the pyserial ones.
I also haven't looked at the code in the pyserial timeouts, don't know if it's actually more efficient at all.

Reply:
We don't want to exit early if the desired pattern is found, because we still need to read and process all incoming chars until it's "quiet", to know it's ready for the remount command.

And I don't think the pyserial timeout settings are enough for the case of 3 different timeout values used here (initial wait, inter-char waits, total timeout).